io/directory.go

```diff
@@ -311,6 +315,11 @@ func (d *HAMTDirectory) SetCidBuilder(builder cid.Builder) {
 
 // AddChild implements the `Directory` interface.
 func (d *HAMTDirectory) AddChild(ctx context.Context, name string, nd ipld.Node) error {
```

Comment:
The switching logic should happen in the Add/Remove child functions, not in the Create/Load function.

If for some reason we are unable to figure out if adding/removing the child will result in needing to switch the directory type I think it's safe to just return an error. We're doing this under the assumption that users would prefer consistency in how an FS directory maps to a UnixFS directory.